src/kv/idbstore.rs

```diff
@@ -145,3 +155,214 @@ impl Store for IdbStore {
+    async fn has(&self, key: &str) -> Result<bool> {
+        let request = self.store.count_with_key(&key.into())?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let callback = Closure::once(move || {
+            if let Err(_) = sender.send(()) {
+                warn!("oneshot send failed");
+            }
+        });
+        request.set_onsuccess(Some(callback.as_ref().unchecked_ref()));
+        request.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        receiver.await?;
+        Ok(match request.result()?.as_f64() {
+            Some(v) if v >= 1.0 => true,
+            _ => false,
+        })
+    }
+
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        let request = self.store.get(&key.into())?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let callback = Closure::once(move || {
+            if let Err(_) = sender.send(()) {
+                warn!("oneshot send failed");
+            }
+        });
+        request.set_onsuccess(Some(callback.as_ref().unchecked_ref()));
+        request.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        receiver.await?;
+        Ok(match request.result()? {
+            v if v.is_undefined() => None,
+            v => Some(js_sys::Uint8Array::new(&v).to_vec()),
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+enum WriteState {
+    Open,
+    Committed,
+    Aborted,
+    Errored,
+}
+
+struct WriteTransaction {
+    rt: ReadTransaction,
+    pending: Mutex<HashMap<String, Vec<u8>>>,
+    pair: Arc<(Mutex<WriteState>, Condvar)>,
+    callbacks: Vec<Closure<dyn FnMut()>>,
+}
+
+impl WriteTransaction {
+    fn new(idb: &IdbStore) -> Result<WriteTransaction> {
+        let tx = idb
+            .idb
+            .transaction_with_str_and_mode(OBJECT_STORE, web_sys::IdbTransactionMode::Readwrite)?;
+        let mut wt = WriteTransaction {
+            rt: ReadTransaction {
+                store: tx.object_store(OBJECT_STORE)?,
+                tx: tx,
+            },
+            pair: Arc::new((Mutex::new(WriteState::Open), Condvar::new())),
+            pending: Mutex::new(HashMap::new()),
+            callbacks: Vec::with_capacity(3),
+        };
+
+        let tx = &wt.rt.tx;
+        let callback = wt.tx_callback(WriteState::Committed);
+        tx.set_oncomplete(Some(callback.as_ref().unchecked_ref()));
+        wt.callbacks.push(callback);
+
+        let callback = wt.tx_callback(WriteState::Aborted);
+        tx.set_onabort(Some(callback.as_ref().unchecked_ref()));
+        wt.callbacks.push(callback);
+
+        let callback = wt.tx_callback(WriteState::Errored);
+        tx.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        wt.callbacks.push(callback);
+
+        Ok(wt)
+    }
+
+    fn tx_callback(&self, new_state: WriteState) -> Closure<dyn FnMut()> {
+        let pair = self.pair.clone();
+        Closure::once(move || {
+            spawn_local(async move {
```

Comment:
Good question! We actually don't need it to happen asynchronously, but we do need to get into an async context to await the mutex. I believe the mutex and condvar must be async because we later wish to block on them in an async context, even though we're in a sync context here in the callback.

I changed spawn_local() to instead be task::block_on(), to indicate those no asynchrony is needed.

I'm not 100% sure why Arc is required, but it's present in all of the CondVar examples, sync and non:
https://doc.rust-lang.org/beta/std/sync/struct.Condvar.html, https://docs.rs/async-std/1.6.2/async_std/sync/struct.Condvar.html. I think it's there to support the `.clone()` operation, which the compiler complains about if it's not present, probably because of the `move`s here. I'm not sure if this is protecting the Mutex or CondVar themselves, or just the anonymous struct that holds the two of them, or what, but I stuck with the examples.